### PR TITLE
[stdlib] Improve default `write_to` function implementations

### PR DESCRIPTION
**Pull request by martinvuyk**
Added tests for previously undiscovered bugs: #5871, #5872, #5873.

This should already make several types only need to use the default implementation for #5870

Changes in behavior:
- Slightly change the behavior of the default `write_to` implementation making it not print the name of the fields
- Make the default skip mlir fields
- SIMD now prints e.g. `SIMD[uint8, 2]([1, 2])`
- InlineArray now prints as a list `[1, 2]` and using repr `InlineArray[Int, 2]([1, 2])`

What is missing:
- A proper way to iterate over parameters and getting their write_repr_to/write_to behavior depending on which one was triggered
- A way to have the Writable interface trickle down the information of whether to use write_repr_to or write_to for nested structures, such that we can use it in all colection types in a generic way and provide a default implementation that does this. It would allow types to implement `write_to` only once

### CODE REVIEW

File: mojo/stdlib/std/format/__init__.mojo [hunk 1]
```diff
@@ -247,31 +240,33 @@ trait Writable:
...
-](this: T, mut writer: W,):
+    T: Writable, //, *, write_repr: Bool = False
+](this: T, mut writer: Some[Writer]):
     comptime names = struct_field_names[T]()
     comptime types = struct_field_types[T]()
     comptime type_name = _unqualified_type_name[T]()
     writer.write_string(type_name)
     writer.write_string("(")
 
+    fn _first_non_private() -> Int:
```
Comment (NathanSWard) [inline mojo/stdlib/std/format/__init__.mojo]: I'm going to push back against not printing "private" fields.
I understand why this would be desirable for the `write_to` method, however Mojo doesn't have an official stance of what "private" fields are (since they don't exist). We just have the current practice of the `_` prefix denoting a private field.

I think we should wait until we officially have these and what their defined syntax before implementing this. :)

If someone really doesn't want their fields with `_` printed, they can simply add a custom `write_to` method.

File: mojo/stdlib/std/format/__init__.mojo [hunk 2]
```diff
@@ -281,12 +276,20 @@ fn _reflection_write_to[
         ]()
 
         @parameter
-        if i > 0:
+        if i > _first_non_private():
             writer.write_string(", ")
-        writer.write_string(materialize[names[i]]())
```
Comment (NathanSWard) [inline mojo/stdlib/std/format/__init__.mojo:286]: I'm also not totally sure on not printing the name of the fields in the default `write_to` method. I'm personally in favor of having the name as it can help differentiate between fields of the same type. Though I can also understand why if someone is trying to debug, they should probably just use `repr`.

File: mojo/stdlib/std/reflection/type_info.mojo [hunk 3]
```diff
@@ -114,18 +114,26 @@ fn get_type_name[
     return StaticString(res)
 
 
-# TODO: This currently does not strip the module name from the inner type name.
-# For example, Generic[Foo] should return "Generic[Foo]" but currently returns
-# "Generic[module_name.Foo]".
-fn _unqualified_type_name[type: AnyType]() -> StaticString:
+fn _unqualified_type_name[type: AnyType]() -> String:
```
Comment (NathanSWard) [inline mojo/stdlib/std/reflection/type_info.mojo]: Having this return a `String` now means any invocation of this will likely allocate. We ideally want to not allocate by returning a `StaticString` or by providing a writer to this function.

Comment (NathanSWard) [inline mojo/stdlib/std/reflection/type_info.mojo]: I'm not sure the implementation of this is fully correct.
For examples cases like `module.Foo[other.Bar]`, and nested generics `Tuple[Tuple[Int, Bool], String]` I don't think work correctly with implementation.

If we do update this, can you please add some unit tests to verify the logic? 
I have a feeling this implementation may be a bit complex, so we it'll likely be nice to pull this out into it's own PR so we can ensure the edge cases are covered :)

**Review by NathanSWard — CHANGES_REQUESTED** [review]

**Comment by NathanSWard** [comment]
Ahh apologies if I jumped on gun on reviewing this, just realized it was a Draft...

File: mojo/stdlib/std/format/__init__.mojo (hunk 1 above)
Comment (martinvuyk) [reply mojo/stdlib/std/format/__init__.mojo]: A lot of our stdlib types have mlir values, so this would be way more tricky to implement for them by default. The point of having a default that is overwritable is for people to implement their own version if they're not happy with it. I would argue having the private fields not printed is an acceptable default that applies in more cases than not

File: mojo/stdlib/std/format/__init__.mojo (hunk 2 above)
Comment (martinvuyk) [reply mojo/stdlib/std/format/__init__.mojo:286]: > Though I can also understand why if someone is trying to debug, they should probably just use repr.

Yep, I was thinking more along those lines and reducing visual noise for the default `write_to` version

File: mojo/stdlib/std/reflection/type_info.mojo (hunk 3 above)
Comment (martinvuyk) [reply mojo/stdlib/std/reflection/type_info.mojo]: What is complicated is adding the logic for removing the fully qualified name from the inner parameters without using `String`, since every for loop iteration would create a new `StaticString`. I can try to do it again, I changed this previously because I found a bug for `Int` that I was trying to work around by using `DevicePassable.get_type_name()` which returns a `String`.

Having this function take in a writer would be very similar to what I'm trying to achieve by adding an interface for the Writer trait to be able to use functions and not just structs (#4793). It could also be solved by allowing a Writer to use a capturing function (#5274)

Comment (martinvuyk) [reply mojo/stdlib/std/reflection/type_info.mojo]: Yeah I think you're right, we could leave this for another time. I think the implementation I did here would be able to deal only with 1 layer of parameter nesting. It will probable need to be recursive which always requires extensive testing :)

**Comment by martinvuyk** [comment]
> Ahh apologies if I jumped on gun on reviewing this, just realized it was a Draft...

Don't worry I did start it off as a review-ready PR, but it seems I have more bugs to isolate besides the ones I found. It could also be that now that I'll remove the `_unqualified_type_name` changes it all just works :) :crossed_fingers: (update: it worked :D)

File: mojo/stdlib/std/format/__init__.mojo (hunk 1 above)
Comment (NathanSWard) [reply mojo/stdlib/std/format/__init__.mojo]: The workaround for dealing with MLIR types should be for us to use `is_struct_type` inside the reflection implementation for `Writable` and then either erroring out or printing `<unprintable>` or something like that.

Until mojo has real "public/private" fields, we should just print it all.
Especially if in the future the syntax is something like this:
```
struct Foo:
    private var x: Int
```

Relying on the stringified name of the field is not a good long term solution IMO.

Comment (martinvuyk) [reply mojo/stdlib/std/format/__init__.mojo]: Ok how about this? We only print fields (including private named ones) if the type doesn't include a flag `_write_self_not_fields` set to true, and those types just call `this.write_to(writer)` instead of writing the fields (when calling repr)? This would help us unify all types like `UnsafePointer` and `Int` in one stroke

Comment (NathanSWard) [reply mojo/stdlib/std/format/__init__.mojo]: > _write_self_not_fields

I still think this is over engineering a bit and bloating the `Writable` trait.
I generally think it is _okay_ for us to have to write custom `write_(repr_)to` implementations for stdlib types, especially those the use raw `__mlir_type`s.
If something has a `__mlir_type` then we should be expected to write a `write_to` method.

I'm not sure exactly how this would unify `UnsafePointer` & `Int` as they both have different MLIR types under the hood and need to be written differently: `UnsafePointer`: the address, `Int`: the numeric value.

I appreciate all of the ideas none the less - it's good to brainstorm 😄. Though I'd like to make small incremental changes, so for this PR I'm open to merging the "not printing field names" in `write_to`, and that would be a good first start.

Comment (martinvuyk) [reply mojo/stdlib/std/format/__init__.mojo]: > I'm not sure exactly how this would unify UnsafePointer & Int as they both have different MLIR types under the hood and need to be written differently: UnsafePointer: the address, Int: the numeric value.

I just did it, you can see the changes in the PR. This unifies most types except for collection types that need to print the repr of their inner elements

>  If something has a __mlir_type then we should be expected to write a write_to method.

That's an alternative but having the flag would allow other custom types to take advantage of their `write_to` having some different representation than their field values

> I still think this is over engineering a bit and bloating the Writable trait.
> I generally think it is okay for us to have to write custom write_(repr_)to implementations for stdlib types, especially those the use raw __mlir_types.

I incorporated the changes and it allowed me to delete a lot of boilerplate and several different types' implementations since they followed the same pattern. Most of the code I've added is a workaround for an issue when printing parameter names, and editing tests for the new expected returns

**Comment by setsource** [comment]
@martinvuyk, I’ve been following the progress here, spot on work! Your work on standardizing the infrastructure with `_write_self_not_fields` is exactly the architectural direction we are requesting for the `stdlib`. However, to fully close out the "Prettier" requirement in **#5838**, I see we have one last gap: **Alias Resolution.**

Currently, code changes print `SIMD[DType.uindex, 1]` for scalars. To satisfy the original request while respecting @NathanSWard’s preference for keeping the `Writable` trait lean, I make a humble proposal of dynamically resolving the `_type_name` within the `SIMD` struct itself. @martinvuyk please consider reviewing my suggestions.

---

- Proposed Refinement for `mojo/stdlib/std/builtin/simd.mojo`

By moving from a static string to a `comptime` helper, we give the reflection engine the "Pretty" name it needs without adding any complexity to the `Writable` trait logic.

**Current in PR:**

```mojo
comptime _type_name = String("SIMD[", Self.dtype, ", ", Self.size, "]")
comptime _write_self_not_fields = True

```

**Proposed Change:**

```mojo
# Inside struct SIMD:

@always_inline
@parameter
fn _get_type_name() -> String:
    @parameter
    if Self.size == 1:
        # Resolve to standard Mojo aliases for visual clarity
        if Self.dtype == DType.int64: return "Int64"
        if Self.dtype == DType.int32: return "Int32"
        if Self.dtype == DType.uint64: return "UInt64"
        if Self.dtype == DType.uint32: return "UInt32"
        if Self.dtype == DType.float64: return "Float64"
        if Self.dtype == DType.index: return "Int"
        if Self.dtype == DType.uindex: return "UInt"
        if Self.dtype == DType.bool: return "Bool"
        return String("Scalar[") + str(Self.dtype) + "]"
    
    return String("SIMD[") + str(Self.dtype) + ", " + str(Self.size) + "]"

comptime _type_name = Self._get_type_name()
comptime _write_self_not_fields = True

```

##### Why these suggestions:

* **Zero Over-Engineering:** As per @NathanSWard suggestions, we aren't changing the trait API or adding runtime overhead. Instead now we are simply providing a smarter `comptime` constant.
* **Satisfies #5838:** That `List[UInt]` will finally print as `[UInt(1), UInt(2)]` instead of the verbose details SIMD internal representation (prettier..)
* **Boilerplate Reduction:** We keep @martinvuyk's excellent work + logic for pruning the `stdlib` of redundant `write_repr_to` implementations.

what do you both think? If we're aligned on this, I hope this will unblock the merge and finally give us those cleaner debug logs we're after. Thanks.

**Comment by martinvuyk** [comment]
@setsource thanks for the proposal, I was actually thinking on doing something way simpler:
```mojo
comptime _upper_idx = 1 + Int(Self.dtype.is_unsigned())
comptime _type_name = String("SIMD[", Self.dtype, ", ", Self.size, "]") if Self.size > 1 else (
    String(Self.dtype)[:Self._upper_idx].upper() + String(Self.dtype)[Self._upper_idx:]
)
```
I didn't implement this in this PR to reduce the scope and try to fix the issue in a separate one, but I could introduce it here if @NathanSWard approves :man_shrugging:

**Comment by NathanSWard** [comment]
@martinvuyk - I don't believe the path of using a `comptime _write_self_not_fields: Bool` and the `compete _type_name` flags is the best approach here. The general idea is that the default reflection-based `write_to` implementation should be a "best guess" at how a type would want to format itself. If for some reason, the type doesn't like the default output it should provide a manual implementation.

This begins to bloat the surface of `Writable` which we should keep simple and make it work for "most cases".
A lot of the stdlib types are not these "most cases" - e.g. custom collections that abstract over pointers, or types that abstract over raw `MLIR` types. These are the exact types that _should_ provide a custom `write_(repr_)to` implementation.

The idea is we should be providing nice "format" utilities. This is was I was trying to achieve with the `FormatStruct/Named` etc... helper types that allow you to compose and build a nice formatter without having to worry about the nuances of adding appropriate whitespace or adding an `=` in the output.

**Comment by martinvuyk** [comment]
@NathanSWard the `_type_name` was actually added as a workaround because the builtin `get_type_name` has issues (https://github.com/modular/modular/issues/5871, https://github.com/modular/modular/issues/5872, https://github.com/modular/modular/issues/5873), it shouldn't be needed to be overwritten for most types except those like SIMD which would like to change their representation based on some parameters.

This is intended to replace a lot of repetitive boilerplate, I don't see it as increasing the surface area by much (It's why I put the `_type_name` as a private alias). I'm just making the default implementation more flexible. I haven't looked into removing recently added custom implementations, but my guess is this will replace a lot more code than what I replaced the last time I rebased on top of the tree.

**Comment by NathanSWard** [comment]
That all makes sense! Though I would instead push for fixing what is currently broken e.g. `_unqualified_type_name` rather than coming up with a temporary work around which is ideally not even needed if `_unqualified_type_name` is correctly fixed.

**Comment by martinvuyk** [comment]
> That all makes sense! Though I would instead push for fixing what is currently broken e.g. `_unqualified_type_name` rather than coming up with a temporary work around which is ideally not even needed if `_unqualified_type_name` is correctly fixed.

@NathanSWard The thing is we will need for types like SIMD to be able to specify their own naming based on parameters if they so wish. I would like to eventually make `_type_name` public, but that decision can wait until post 1.0 . This is intended as a workaround/maybe future feature. I have use-cases for custom type naming (without having to write a bunch of boilerplate) easily come to mind for other community projects, I think this feature will be nice.

FWIW I tried patching just `_unqualified_type_name` by having `AnyType` have an alias with the `get_type_name` result but everything exploded, I also tried other approaches but it won't work because the `get_type_name` function returns are just a mess. Also, I've reported these bugs a month ago and I don't think they'll be quickly fixed. And the concern of stringifying a type name does seem to me like the responsibility of the `Writable` trait.
